src/util/address.rs

```diff
@@ -139,6 +139,7 @@ impl From<bech32::Error> for Error {
 
 /// The different types of addresses.
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[non_exhaustive]
```

Comment:
Thanks for the review. I'm not totally sure of the merit of this PR myself, whether we should use `non_exhaustive`, or whether we should use it but on some different set of types. There has been multiple folks speak up in the past positively about `non_exhaustive`, the intention of this PR is to bring out all the opinions and then I'll just try to shepherd them in. As such, I'll leave this sitting open for a while and see what comes of it.

Reply:
There are much better ways to detect whether an address is witness (or something else) than matching each type. I think bumping the major version just because there's a new *soft* fork is annoying. The issue with this library is that pretty much any Bitcoin application written in Rust uses it, so it's a shitton of breakage.

IMO any public enum and struct that could be possibly extended in the future should be `#[non_exhaustive]`

Reply:
I tend to agree with @Kixunil here ... my feeling is that matching on these enums is a bit of a weird thing to do, and if you are doing this, you ought to have an error path for the case that some unsupported softfork has happened. (Or in some cases perhaps you could do some sane default.)

On the other hand, soft forks in Bitcoin *are* really rare, so this is something of an academic discussion. Maybe the time to discuss this is whenever the next fork actually seems to be happening? (My guess is that the next fork will be addition of some covenant opcodes, which won't even bump any version numbers, though it could also be APO..)

Reply:
The problem is it's not "just updating a few enums". Due to breaking change there needs to be a coordinated major version bump across all ecosystem. This is a huge library and it is the only (rich) Bitcoin library available in Rust today, so pretty much every Bitcoin project written in Rust depends on it. Perhaps if `Address` was a separate crate it'd be less painful but I guess it'd still be significant.

This is my main motivation behind trying to make the library stable as you might have noticed from my efforts.

> Maybe the time to discuss this is whenever the next fork actually seems to be happening?

I don't think so, if at that time we have 1.0 it's already too late. At the latest we need to decide before making it 1.0.